Initial commit for atSpoke integration

kevincasey1222 opened the pull request:
This integration uses Axios and the Jupiter One integration template. 

It gathers 6 types of objects - Account, Users, Teams (J1 UserGroup), Requests (J1 Record), RequestTypes (J1 Configuration), and Webhooks (J1 ApplicationEndpoint). All but Webhooks support pagination (the atSpoke API doesn't support pagination for GET /webhooks, but these are outbound APIs - there shouldn't be that many).

The config file asks for an API_KEY and NUM_REQUESTS param which sets the number of Requests (essentially helpdesk tickets) to pull into the system, all statuses. The atSpoke API returns the most recent X numbers of requests, set by this param. These Records in J1 are set with partial:true so they can accumulate. Optionally, you can set NUM_REQUESTS to 0, in which case the integration will not pull Requests at all, nor will it pull RequestTypes. (RequestTypes are also set partial:true, so no problem if old entries are in the system).

Tests use a recording and compare it to a snapshot, both drawn from a test account I made on atSpoke.

kevincasey1222 committed:
Initial commit

kevincasey1222 committed:
Updated requests and requestTypes to partial

kevincasey1222 committed:
Added pagination

kevincasey1222 committed:
Added pagination for requests, request types, users, teams

kevincasey1222 committed:
Updated snapshots, request status, and a property name typo

kevincasey1222 commented:
I'm confused as to why the build test is saying that the snapshots don't match. I've run "yarn run prepush -u" and "yarn test -u" here. Also, all the tests pass when I push. What is different about the CICD environment?

kevincasey1222 committed:
Snapshot update

kevincasey1222 commented:
Alright, snapshot solved. Not sure why that needed separate wrangling, but it is wrangled.

aiwilliams commented on README.md:
```diff
@@ -1,50 +1,104 @@
-with JupiterOne in the [integration documentation](docs/jupiterone.md).
-
-## Development
+## Development Environment
```
@kevincasey1222 the integration template was changed between the time that you first started this work and the time this PR was created (I think). We'd like to have the README.md the same on all integrations, which will point to other documents that describe how integrations work, and any unique things will be in the `docs/` directory. I think that means:

1. This file can simply be reverted.
2. "Getting an API Key from atSpoke" and "Configuring the number of requests desired in your JupiterOne graph" in here can be moved to `docs/development.md`.

kevincasey1222 replied:
Done. Thanks for the clarity.

aiwilliams commented on src/types.ts:
```diff
@@ -6,12 +6,8 @@ import { IntegrationInstanceConfig } from '@jupiterone/integration-sdk-core';
-   */
-  clientSecret: string;
+  apiKey: string;
+  numRequests: string;
```
It will be important to document the meaning and impact of changing this variable. We are going to make it visible in the UI as well (when we deploy it), so this will aid writing the help text for the field.

```suggestion
  /**
   * The number of recent requests to ingest.
   *
   * Requests are Record entities. The collection size will grow indefinitely. This allows for limiting the number
   * of recently created/changed requests to ingest whenever the integration executes. This should be set to
   * a value that correlates to the execution interval of the integration and the expected number of requests
   * that would have been created/changed between executions.
   */
  numRequests: string;
```

aiwilliams commented on src/validateInvocation.test.ts:
```diff
@@ -32,8 +32,8 @@ it('auth error', async () => {
     instanceConfig: {
-      clientId: 'INVALID',
-      clientSecret: 'INVALID',
+      apiKey: 'INVALID',
+      numRequests: '0',
```
It would be good to test what happens when `numRequests` is `undefined`, `-1000`, `abc`, etc. since this will be configured by users.

kevincasey1222 replied:
Many tests added to parser. Thanks!

aiwilliams commented on src/steps/index.test.ts:
```diff
@@ -79,20 +100,35 @@ test('should collect data', async () => {
+  const webhooks = context.jobState.collectedEntities.filter((e) =>
+    e._class.includes('ApplicationEndpoint'),
+  );
+  //expect(webhooks.length).toBeGreaterThan(0);
```
I think it would be good to make this say we ingested at least one. This will mean the development/test account would have to have one setup of course, but that's a good thing should someone need to re-record responses.

kevincasey1222 replied:
Done.

aiwilliams commented on src/steps/account.ts:
```diff
@@ -5,27 +5,31 @@ import {
+import { createAPIClient } from '../client';
 
-export const ACCOUNT_ENTITY_KEY = 'entity:account';
+export const ACCOUNT_ENTITY_KEY = 'at_spoke_account';
```
I'm going to update the template project....

We're working toward using this naming pattern for the constant that defines the key for `setData/getData` entries, to clarify that this is meant to be orthogonal to entities and their `_type` values. Then, all things stuck in `setData/getData` will be easily found by `DATA_` search.

```suggestion
export const DATA_ACCOUNT_ENTITY = 'DATA_ACCOUNT_ENTITY';
```

kevincasey1222 replied:
Implemented.

aiwilliams commented on src/steps/account.ts:
```diff
@@ -5,27 +5,31 @@ import {
-          _key: 'acme-unique-account-id',
-          _type: 'acme_account',
+          _key: `at-spoke-account:${instance.id}`,
+          _type: ACCOUNT_ENTITY_KEY,
```
With the aforementioned move to `DATA_ACCOUNT_ENTITY`, it will become clear that this should be a different variable with the value `at_spoke_account`.

This is not necessary to address in this PR, but wanted you to be aware, and I've got a task to update the integration-template...

We're moving toward this pattern for defining the things ingested and referencing them in the step metadata: https://github.com/JupiterOne/graph-azure/blob/master/src/steps/resource-manager/network/constants.ts

kevincasey1222 replied:
Implemented.

aiwilliams commented on src/steps/account.ts:
```diff
@@ -5,27 +5,31 @@ import {
-          _key: 'acme-unique-account-id',
-          _type: 'acme_account',
+          _key: `at-spoke-account:${instance.id}`,
+          _type: ACCOUNT_ENTITY_KEY,
```
With the aforementioned move to `DATA_ACCOUNT_ENTITY`, it will become clear that this should be a different variable with the value `at_spoke_account`.

This is not necessary to address in this PR, but wanted you to be aware, and I've got a task to update the integration-template...

We're moving toward this pattern for defining the things ingested and referencing them in the step metadata: https://github.com/JupiterOne/graph-azure/blob/master/src/steps/resource-manager/network/constants.ts

kevincasey1222 replied:
Implemented

aiwilliams commented on src/steps/requests.ts:
```diff
@@ -0,0 +1,169 @@
+            _type: 'at_spoke_request',
+            _class: 'Record',
+            _key: request.id,
+            partial: true,
```
This is a value that should be on the metadata in the step.

aiwilliams commented on src/steps/requests.ts:
```diff
@@ -0,0 +1,169 @@
+            _type: 'at_spoke_requesttype',
+            _class: 'Configuration',
+            _key: requestType.id,
+            partial: true,
```
This is a value that should be on the metadata in the step.

aiwilliams commented on src/steps/requests.ts:
```diff
@@ -0,0 +1,169 @@
+        _type: 'at_spoke_account',
+        _class: 'Account',
+      },
+      {
+        resourceName: 'atSpoke Request',
+        _type: 'at_spoke_request',
+        _class: 'Record',
+      },
```
Here is where the `partial: true` value would go.

Note that the relationships between the account and request would also need to be marked as `partial: true`. This flag is metadata informing the synchronization system that there is no need to do set comparison (leading to deletes of objects which are in the graph but not represented in the newly uploaded set), which is true for these relationships because the integration will never upload a complete set, just as it does not for the request Record entities.

kevincasey1222 replied:
Thanks, @aiwilliams . So, when I got to add partial: true here, it tells me that partial does not exist as a known property in StepEntityMetadata.ts.

aiwilliams commented on src/steps/__recordings__/steps_741716276/recording.har:
```diff
@@ -0,0 +1,740 @@
+              "value": "application/json"
+            },
+            {
+              "name": "api-key",
+              "value": "AtFbNFj2hvzt7pGkg_zgW7ksihiFaVrVwttLKxU2oRVU="
```
The recording setup code should add this as a sensitive header key so that the value is redacted from the recordings. Also, it is now necessary to delete this API key from the provider and generate a new one.

kevincasey1222 replied:
Done.

aiwilliams commented on src/steps/__recordings__/steps_741716276/recording.har:
```diff
@@ -0,0 +1,740 @@
+          "content": {
+            "mimeType": "application/json; charset=utf-8",
+            "size": 408,
+            "text": "{\"results\":[{\"authentication\":{\"type\":\"client_secret\",\"clientSecret\":\"EhKHiEzCCbd7wSRyByU1TISyAeuS8nEh3ThwGqC2h2c=\"},\"enabled\":true,\"topics\":[\"request.created\"],\"url\":\"https://example.com/receive_hook\",\"client\":\"Zapier\",\"description\":\"Push resource creation to external system.\",\"createdAt\":\"2021-02-12T22:10:34.105Z\",\"org\":\"jupiteronedev\", ...
```
This may need some redaction work since it contains a secret looking thing. When we don't redact these, even if they're not terribly sensitive because these are dev accounts, we end up getting hackers saying that we have a leak, and we have to treat it as super important, which can be a hassle/unnecessary scare. That said...

Is this a value that can be configured when setting up the webhook? Perhaps it could be set to something like "not-real-very-not-sensitive-actually"?

kevincasey1222 replied:
Done.

kevincasey1222 replied:
And yes, the only way to make a webhook is via REST API, so I just did that once in the dev acct so I could pull it down. In this case, I just updated it in the recording and snapshot.

aiwilliams commented on src/steps/__snapshots__/index.test.ts.snap:
```diff
@@ -7,141 +7,1133 @@ Object {
+          "role": "MEMBER",
+          "status": "INVITED",
+        },
+      ],
```
So, I'm sorry about this, we need to do something in the SDK to prevent this... just created https://github.com/JupiterOne/sdk/issues/424.

Graph objects (entities, relationships) cannot have properties that are objects, nor Arrays of objects. The synchronization system will reject this as an invalid entity.

This property can, I think, simply be dropped at this time. Perhaps we could have another change later on where this drives a relationship to each organization entity the user has a membership in, so that an analyst can see where their users are accessing multiple orgs with their user accounts?

kevincasey1222 replied:
Property dropped.

aiwilliams commented on src/steps/__snapshots__/index.test.ts.snap:
```diff
@@ -7,141 +7,1133 @@ Object {
+          "name": "default",
+          "rawData": Object {
+            "_id": "601da2f204e2850006039574",
+            "agentList": Array [
```
Let's drop `agentList` from the raw data. These should be the raw data of the entities themselves.

kevincasey1222 replied:
Done.

aiwilliams commented on src/steps/__snapshots__/index.test.ts.snap:
```diff
@@ -7,141 +7,1133 @@ Object {
+            "slug": "test-team-alpha",
+            "status": "ACTIVE",
+            "updatedAt": "2021-02-05T19:56:34.846Z",
+            "users": Array [
```
Let's drop `users` from the raw data. These should be the raw data of the user entities.

kevincasey1222 replied:
Done.

aiwilliams commented on src/steps/__snapshots__/index.test.ts.snap:
```diff
@@ -7,141 +7,1133 @@ Object {
+          "name": "default",
+          "rawData": Object {
+            "authentication": Object {
+              "clientSecret": "EhKHiEzCCbd7wSRyByU1TISyAeuS8nEh3ThwGqC2h2c=",
```
This will need to be redacted from the raw data.

kevincasey1222 replied:
Done.

aiwilliams commented on src/steps/__snapshots__/index.test.ts.snap:
```diff
@@ -7,141 +7,1133 @@ Object {
+            "permalink": "https://jupiteronedev.askspoke.com/next/requests/4",
+            "privacyLevel": "public",
+            "requestType": "6026f670ac67a90007a282f1",
+            "requestTypeInfo": Object {
```
I think we should drop this from this raw data. It should be the raw data of the request types themselves. Otherwise, we'll repeat all this throughout every request entity.

Seems like their API is generous with the data it returns ATM...

kevincasey1222 replied:
Done.

aiwilliams commented on src/steps/__snapshots__/index.test.ts.snap:
```diff
@@ -7,141 +7,1133 @@ Object {
+      "displayName": "Request for “Broken hardware”",
+      "email": "request-6026f691ae0b320006140f3e@jupiteronedev.askspoke.com",
+      "name": "Request for “Broken hardware”",
+      "owner": "602585518843b90007eeda34",
```
Let's create an Issue to build a relationship to owner user entities.

aiwilliams commented on src/steps/__snapshots__/index.test.ts.snap:
```diff
@@ -7,141 +7,1133 @@ Object {
+      "owner": "602585518843b90007eeda34",
+      "partial": true,
+      "requestType": "6026f670ac67a90007a282f1",
+      "requester": "601da2951d25d80007bf87d2",
```
Let's create an Issue to build a relationship to requester user entities.

aiwilliams commented on src/steps/__snapshots__/index.test.ts.snap:
```diff
@@ -7,141 +7,1133 @@ Object {
+            "requester": "601da2951d25d80007bf87d2",
+            "status": "OPEN",
+            "subject": "Test msg for Slack integration",
+            "tags": Array [],
```
Let's add an Issue to convert these to `tag.<name>` properties on the entity.

aiwilliams commented on src/steps/access.ts:
```diff
@@ -107,32 +127,68 @@ export const accessSteps: IntegrationStep<IntegrationConfig>[] = [
-        resourceName: 'Account',
-        _type: 'acme_account',
+        resourceName: 'atSpoke Account',
+        _type: 'at_spoke_account',
```
The `fetch-account` step creates this entity, so it can be removed from this step's metadata.

aiwilliams commented on src/steps/access.ts:
```diff
@@ -107,32 +127,68 @@ export const accessSteps: IntegrationStep<IntegrationConfig>[] = [
-        _type: 'acme_account_has_group',
+        resourceName: 'atSpoke Account',
+        _type: 'at_spoke_account',
+        _class: 'Account',
```
The `fetch-account` step creates this entity, so it can be removed from this step's metadata.

aiwilliams commented on src/steps/access.ts:
```diff
@@ -107,32 +127,68 @@ export const accessSteps: IntegrationStep<IntegrationConfig>[] = [
+        resourceName: 'atSpoke User',
+        _type: 'at_spoke_user',
+        _class: 'User',
+      },
```
The `fetch-users` step creates this entity, so it can be removed from this step's metadata.

aiwilliams commented on src/steps/access.ts:
```diff
@@ -107,32 +127,68 @@ export const accessSteps: IntegrationStep<IntegrationConfig>[] = [
+  },
+  {
+    id: 'fetch-groups',
+    name: 'Fetch UserGroups',
```
I *think* these names tend to use the terminology of the provider, so that in the job logs presented to the users, they see that the atSpoke Teams have been processed.
```suggestion
    id: 'fetch-teams',
    name: 'Fetch Teams',
```

kevincasey1222 replied:
True. Done.

kevincasey1222 replied:
Agree. Done.

aiwilliams commented on src/steps/access.ts:
```diff
@@ -107,32 +127,68 @@ export const accessSteps: IntegrationStep<IntegrationConfig>[] = [
+        _class: RelationshipClass.HAS,
+        sourceType: 'at_spoke_account',
+        targetType: 'at_spoke_user',
+      },
```
The `fetch-users` step creates this relationship, so it can be removed from this step's metadata.

aiwilliams commented on src/steps/requests.ts:
```diff
@@ -0,0 +1,169 @@
+        resourceName: 'atSpoke Account',
+        _type: 'at_spoke_account',
+        _class: 'Account',
+      },
```
The `fetch-account` step creates this entity, so it can be removed from this step's metadata.

aiwilliams commented on src/steps/requests.ts:
```diff
@@ -0,0 +1,169 @@
+    id: 'fetch-request-types',
+    name: 'Fetch Request Types',
+    entities: [
+      {
+        resourceName: 'atSpoke Account',
+        _type: 'at_spoke_account',
+        _class: 'Account',
+      },
```
The `fetch-account` step creates this entity, so it can be removed from this step's metadata.

aiwilliams commented on src/steps/webhooks.ts:
```diff
@@ -0,0 +1,82 @@
+    id: 'fetch-webhooks',
+    name: 'Fetch Webhooks',
+    entities: [
+      {
+        resourceName: 'atSpoke Account',
+        _type: 'at_spoke_account',
+        _class: 'Account',
+      },
```
The `fetch-account` step creates this entity, so it can be removed from this step's metadata.

aiwilliams reviewed (changes requested):
Thanks @kevincasey1222! I made some inline comments, but these ones here couldn't be attached to a file. I think most of this will be pretty quick to fix.

1. [`docs/jupiterone.md`](https://github.com/JupiterOne/graph-atspoke/pull/1/files#diff-5aa6cbd3ecfc78d03ab6ffe94dac2c4ff3081fd66f15b999d7691ddad293b547) has a number of `TODO` statements in it that need to be addressed. Consider the Google integration as [an example](https://github.com/JupiterOne/graph-google/blob/master/docs/jupiterone.md).
2. We currently use Yarn for package installation. I see there is a [`package-lock.json`](https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json), which is produced by NPM. I think the latest version of NPM can use a `yarn.lock` file, but I [don't think it works the other way](https://classic.yarnpkg.com/blog/2018/06/04/yarn-import-package-lock/).

kevincasey1222 committed:
Tests updated

kevincasey1222 committed:
Merge branch 'initial' of github.com:kevincasey1222/graph-atspoke into initial

kevincasey1222 committed:
Updated recordings to redact

kevincasey1222 committed:
Removed memberships array from users

kevincasey1222 committed:
Removed arrays and objects from graph rawdata

kevincasey1222 committed:
Updated metadata and changed all group refs to teams

kevincasey1222 committed:
More metadata cleanup

kevincasey1222 committed:
Yet more metadata cleanup.

kevincasey1222 commented:
Thanks for the detailed code review, @aiwilliams! I've fixed most of the items, made a few improvements here and there, and created issues for the three where you recommended that.

The only thing not addressed is the partial: true. You said that shouldn't be an entity property but rather a metadata entry? But when I try to add it as a metadata entry it says that the property `partial` is not a valid option in StepEntityMetadata.ts.

aiwilliams commented on docs/jupiterone.md:
```diff
@@ -1,33 +1,22 @@
 # Integration with JupiterOne
 
-## {{provider}} + JupiterOne Integration Benefits
+## atSpoke + JupiterOne Integration Benefits
```
Excellent work on this file, thanks a ton!

aiwilliams commented on src/validateInvocation.test.ts:
```diff
@@ -37,7 +38,52 @@ it('auth error', async () => {
+
+//since parseNumRequests will test a user-configurable field, it's best to be thorough
+//expected return is 0 for all cases except where a valid number can be derived
+test('parser test 1', () => {
+  expect(parseNumRequests("0")).toBe(0);
+});
+test('parser test 2', () => {
+  expect(parseNumRequests("some string")).toBe(0);
+});
```
🔥 TIP 🚒 : check out https://jestjs.io/docs/en/api#testeachtablename-fn-timeout

aiwilliams reviewed (commented):
Thanks a ton for addressing all that @kevincasey1222 👍 

Re: `partial`, check out https://github.com/JupiterOne/sdk/pull/371/files#diff-515bb61dc6b28aad92c0bd1f754a5d59808f4603622cc64de09f0814ed94830dR90. Hopefully that clears things up a bit.

It looks like `test/recording.ts` was not changed to automatically redact sensitive things in new recordings. Did you do that manually? That's okay for this PR I think, just want to be sure I didn't miss anything. We can make it automatic when the recordings are changed in the future.

kevincasey1222 committed:
Updated SDK dependency version

kevincasey1222 committed:
Updated recording redaction, fixed request partial flag

kevincasey1222 committed:
Snapshot fix required

kevincasey1222 commented:
Hi @aiwilliams ! 

You're right, I updated the recording manually. I hadn't realized there was a way to automatically redact sensitive headers in the recordings. I updated the invocation to redact `api-key` automatically, which was most of the issue. However, the webhooks have a `client-secret` property in the response body returned by the API when we get webhooks, so those will have to be manually redacted in the future. That said, webhooks can only be created by the API also, so in the future, if we create a webhook in a dev acct, we can just set the client-secret property to "not-really-a-secret" or "redacted_secret" and then we won't have to do any manual updating of the recording.

I also figured out the `partial: true` flag - it was a problem because the integration template was using sdk 3.1.0 by default. I updated to 5.8.0 and all was good. In moving `partial` from being an entity property to a metadata flag, I also had to update the snapshot.

All should be good now.

aiwilliams reviewed (approved):
Thanks @kevincasey1222 !
